### DISCUSSION TITLE
Questions about CLI use-cases w/out compilation

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1050] [Q&A]
I have a bunch of CLI use-cases for which I'm not sure whether there is a solution or not:

1. I have a board running some app, it has a serial adapter attached, I'd like to connect to it and see the `trace()` output printed so I see what it's up to, how can I do that?
2. Same situation as previous bullet, except that I'd like to reset the board and then see what it says.
3. Same situation as first or second bullet, except that I'd like to attach xsbug to troubleshoot what it's doing.
4. I'd like to load a mod onto a board (mcrun) and see the `trace()` output in the terminal

Different but somewhat related question:
- is there a way to find out when the xsbug protocol fails to connect and shuts down (that's what it does, right?), and then take over the serial port to print log lines in plain text?

**Comment by phoddie** [comment]
I'll assume you are using ESP32.

> I have a board running some app, it has a serial adapter attached, I'd like to connect to it and see the trace() output printed so I see what it's up to, how can I do that?

Once the app is running, it is not possible to connect to xsbug. The protocol is initialized at start-up.

> Same situation as previous bullet, except that I'd like to reset the board and then see what it says.

`mcconfig -d -m -p esp32 -l -t xsbug`

Please see the [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) for a note about setting up the script used by `-l`.

> Same situation as first or second bullet, except that I'd like to attach xsbug to troubleshoot what it's doing.

`mcconfig -d -m -p esp32 -t xsbug`

> I'd like to load a mod onto a board (mcrun) and see the trace() output in the terminal

Not currently implemented, but would be straightforward to implementsupport that parallels  mcconfig by updating the make files in `$MODDABLE/tools/mcrun.` Then this would work:

`mcrun -d -m -p esp32 -l`

That would be a valuable PR to contribute.

> is there a way to find out when the xsbug protocol fails to connect and shuts down (that's what it does, right?), and then take over the serial port to print log lines in plain text?

An instrumented build on ESP32 redirects `trace` to the serial output that xsbug isn't using.
